### PR TITLE
[Archetype builder] Make the "archetype anchor" use a total order. 

### PR DESCRIPTION
**Pull request by DougGregor**
Make the "archetype anchor" algorithm use the same total order used to
compare dependent types in a generic signature, so that it is a
dependable total order. Tighten up the `compareDependentTypes()` logic
(and checking of it) a bit now that it's being used more frequently.

The weird code around `GenericSignature::isCanonicalTypeInContext()` is
due to a longstanding issue where

    ArchetypeBuilder::resolveArchetype(t)->getDependentType()

is not the identity function when the incoming type is a
`DependentMemberType` with known associated types. We want to establish
this behavior at some point going forward, in which case we can go
back to the prior implementation of
`GenericSignature::isCanonicalTypeInContext()`.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: test/IRGen/associated_types.swift [hunk 1]
```diff
@@ -76,21 +76,16 @@ func testFastRuncible<T: Runcible, U: FastRuncible where T.RuncerType == U.Runce
...
-// CHECK-NEXT: [[T0:%.*]] = load i8*, i8** %T.RuncerType.Runcer,
-// CHECK-NEXT: [[T1:%.*]] = bitcast i8* [[T0]] to %swift.type* (%swift.type*, i8**)*
-// CHECK-NEXT: %T.RuncerType.Runcee = call %swift.type* [[T1]](%swift.type* %T.RuncerType, i8** %T.RuncerType.Runcer)
 //   2. Get the witness table for U.RuncerType.Runcee : Speedy
 //     2a. Get the protocol witness table for U.RuncerType : FastRuncer.
 // CHECK-NEXT: [[T0:%.*]] = getelementptr inbounds i8*, i8** %U.FastRuncible, i32 1
 // CHECK-NEXT: [[T1:%.*]] = load i8*, i8** [[T0]],
 // CHECK-NEXT: [[T2:%.*]] = bitcast i8* [[T1]] to i8** (%swift.type*, %swift.type*, i8**)*
 // CHECK-NEXT: %T.RuncerType.FastRuncer = call i8** [[T2]](%swift.type* %T.RuncerType, %swift.type* %U, i8** %U.FastRuncible)
+//     1c. Get the type metadata for U.RuncerType.Runcee.
```
Comment (slavapestov) [inline test/IRGen/associated_types.swift]: The runcer becomes the runcee! A life lesson.

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
WAT

**Comment by DougGregor** [comment]
Ah, there's some nondeterminism here because we're not setting up all of the same-type constraints that we need among same-named associated types.

**Comment by DougGregor** [comment]
I merged this separately.

**Pull request closed without merging**